hopsworks-common/src/main/java/io/hops/hopsworks/common/user/UsersController.java

```diff
@@ -288,40 +282,97 @@ public void addOrg(Users user) {
     org.setPhone("-");
     user.setOrganization(org);
   }
-
-  public void recoverPassword(String email, String securityQuestion, String securityAnswer, HttpServletRequest req)
-    throws UserException, ServiceException {
-    if (userValidator.isValidEmail(email) && userValidator.isValidsecurityQA(securityQuestion, securityAnswer)) {
-      Users user = userFacade.findByEmail(email);
-      if (user == null) {
-        throw new UserException(RESTCodes.UserErrorCode.USER_WAS_NOT_FOUND, Level.FINE);
-      }
-      if (!authController.validateSecurityQA(user, securityQuestion, securityAnswer, req)) {
-        throw new UserException(RESTCodes.UserErrorCode.SEC_QA_INCORRECT, Level.FINE);
-      }
-      authController.resetPassword(user, req);
+  
+  public void sendQRRecoveryEmail(String email, String password, HttpServletRequest req)
+    throws UserException, MessagingException {
+    Users user = userFacade.findByEmail(email);
+    String path = FormatUtils.getUserURL(req);
+    if (!authController.checkUserPasswordAndStatus(user, password, req)) {
+      throw new UserException(RESTCodes.UserErrorCode.INCORRECT_CREDENTIALS, Level.FINE);
+    }
+    if (!user.getTwoFactor()) {
+      throw new UserException(RESTCodes.UserErrorCode.TWO_FA_DISABLED, Level.FINE);
     }
+    authController.sendNewRecoveryValidationKey(user, path, false, req);
+  }
+  
+  public void sendPasswordRecoveryEmail(String email, String securityQuestion, String securityAnswer,
+    HttpServletRequest req) throws UserException, MessagingException {
+    Users user = userFacade.findByEmail(email);
+    String path = FormatUtils.getUserURL(req);
+    if (!authController.validateSecurityQAndStatus(user, securityQuestion, securityAnswer, req)) {
+      throw new UserException(RESTCodes.UserErrorCode.SEC_QA_INCORRECT, Level.FINE);
+    }
+    authController.sendNewRecoveryValidationKey(user, path, true, req);
+  }
+  
+  public String recoverQRCode(String key, HttpServletRequest req) throws UserException, MessagingException {
+    return new String(recoverQRCodeByte(key, req));
+  }
+  
+  public byte[] recoverQRCodeByte(String key, HttpServletRequest req) throws UserException, MessagingException {
+    Users user = authController.validateRecoveryKey(key, ValidationKeyType.QR_RESET, req);
+    byte[] qrCode = recoverQRCode(user, req);
+    if (qrCode == null) {
+      throw new UserException(RESTCodes.UserErrorCode.TWO_FA_DISABLED, Level.FINE);
+    }
+    authController.resetValidationKey(user);
+    //Send verification
+    String subject = UserAccountsEmailMessages.ACCOUNT_QR_RESET;
+    String msg = UserAccountsEmailMessages.buildQRResetMessage();
+    emailBean.sendEmail(user.getEmail(), Message.RecipientType.TO, subject, msg);
+    return Base64.encodeBase64(qrCode);
+  }
+  
+  private byte[] recoverQRCode(Users user, HttpServletRequest req) {
+    String random = SecurityUtils.getRandomPassword(Settings.PASSWORD_MIN_LENGTH);
```

Comment:
Why not use `getSecureRandom` here?

Reply:
This code is just moved from https://github.com/logicalclocks/hopsworks/pull/279/files/127647f65e3d2e20a1b7f3a28ebb107b8648b32b#diff-7b18d017a63053a6feb7d4ed10f37140L199